src/cargo/core/dependency.rs

```diff
@@ -303,8 +303,20 @@ impl Dependency {
     }
 
     /// Lock this dependency to depending on the specified package id
-    pub fn lock_to(self, id: &PackageId) -> Dependency {
-        self.clone_inner().lock_to(id).into_dependency()
+    pub fn lock_to(mut self, id: &PackageId) -> Dependency {
+        Rc::make_mut(&mut self.inner).lock_to(id);
+        self
+    }
+
+    /// Set the version requirement for this dependency
+    pub fn set_version_req(&mut self, req: VersionReq) -> &mut Dependency {
+        Rc::make_mut(&mut self.inner).set_version_req(req);
+        self
+    }
+
+    pub fn set_kind(&mut self, kind: Kind) -> &mut Dependency {
```

Comment:
Since the discovery of `Rc::make_mut` I actually think that `DependencyInner` is just no longer needed entirely. The perf should be the same if you just use `&mut self` methods on `Dependency` as they'd all hit the "I'm an  owned Rc" fast path anyway.

I don't mind doing that as part of this!